File: lab1/cpp/main.cpp
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/video.hpp>

using namespace cv;
using namespace std;

void trackPoint(VideoCapture capture, Mat old_gray, vector<Point2f> p0, vector<Point2f> p1, Mat mask, vector<Scalar> colors)
{
    while (true)
    {
        Mat frame, frame_gray;
        capture >> frame;
        if (frame.empty())
        {
            cout << "frame break from here" << endl;
            break;
        }
        cvtColor(frame, frame_gray, COLOR_BGR2GRAY);
        vector<uchar> status;
        vector<float> err;
        TermCriteria criteria = TermCriteria((TermCriteria::COUNT) + (TermCriteria::EPS), 10, 0.03);
        calcOpticalFlowPyrLK(old_gray, frame_gray, p0, p1, status, err, Size(30, 30), 2, criteria);
        vector<Point2f> good_new;
        for (uint i = 0; i < p0.size(); i++)
        {
            if (status[i] == 1)
            {
                good_new.push_back(p1[i]);
                line(mask, p1[i], p0[i], colors[i], 2);
                circle(frame, p1[i], 5, colors[i], -1);
            }
        }
        Mat img;
        cv::add(frame, mask, img);
        cv::imshow("Frame", img);
        int keyboard = waitKey(30);
        if (keyboard == 'q' || keyboard == 27)
            break;
        // Now update the previous frame and previous points
        old_gray = frame_gray.clone();
        p0 = good_new;
    }
}

int main(int argc, char **argv)
{
    const string keys =
        "{ h help |      | print this help message }"
        "{ @image | vtest.avi | path to image file }";
    CommandLineParser parser(argc, argv, keys);
    cout << parser.get<string>("@image") << endl;
    string name = parser.get<string>("@image");
    string filename = "";
    if (name != "vtest.avi")
    {
        filename = samples::findFile(parser.get<string>("@image"));
        if (!parser.check())
        {
            parser.printErrors();
            return 0;
        }
    }

    cv::VideoCapture capture;
    if (filename == "")
    {
        int deviceID = 0;
        int apiID = cv::CAP_ANY;
        capture.open(deviceID, apiID);
    }
    else
    {
        capture.open(filename);
    }

    if (!capture.isOpened())
    {
        cerr << "ERROR! Unable to open camera\n";
        return -1;
    }

    vector<Scalar> colors;
    RNG rng;

    for (int i = 0; i < 100; i++)
    {
        int r = rng.uniform(0, 256);
        int g = rng.uniform(0, 256);
        int b = rng.uniform(0, 256);
        colors.push_back(Scalar(r, g, b));
    }

    Mat old_frame, old_gray;
    vector<Point2f> p0, p1;

    capture >> old_frame;
    cvtColor(old_frame, old_gray, COLOR_BGR2GRAY);
    goodFeaturesToTrack(old_gray, p0, 100, 0.3, 7, Mat(), 7, false, 0.04);
    Mat mask = Mat::zeros(old_frame.size(), old_frame.type());

    try
    {
        trackPoint(capture, old_gray, p0, p1, mask, colors);
    }
    catch (Exception e)
    {
        cout << "caught an exception";
        goodFeaturesToTrack(old_gray, p0, 100, 0.3, 7, Mat(), 7, false, 0.04);
        Mat mask = Mat::zeros(old_frame.size(), old_frame.type());
        trackPoint(capture, old_gray, p0, p1, mask, colors);
    }
}

File: lab1/py/main.py
import numpy as np
import cv2 as cv
import argparse

parser = argparse.ArgumentParser(description="Sample Demo Lucas-Kanade")
parser.add_argument('--video', type=str, help="path to video file")
parser.add_argument('--useCamera', type=bool,
                    help="check this flag to capture from inbuilt camera")

args = parser.parse_args()
cap = None

if(args.useCamera):
    cap = cv.VideoCapture(0)
else:
    cap = cv.VideoCapture(args.video)

feature_params = dict(maxCorners=100,
                      qualityLevel=0.01,
                      minDistance=7,
                      blockSize=7)

lk_params = dict(winSize=(15, 15), maxLevel=2, criteria=(
    cv.TERM_CRITERIA_EPS | cv.TERM_CRITERIA_COUNT, 10, 0.03))

color = np.random.randint(0, 255, (100, 3))

ret, old_frame = cap.read()
old_gray = cv.cvtColor(old_frame, cv.COLOR_BGR2GRAY)
p0 = cv.goodFeaturesToTrack(old_gray, mask=None, **feature_params)

mask = np.zeros_like(old_frame)

while(1):
    ret, frame = cap.read()
    frame_gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)

    p1, st, err = cv.calcOpticalFlowPyrLK(
        old_gray, frame_gray, p0, None, **lk_params)
    if p1 is not None:
        good_new = p1[st == 1]
        good_old = p0[st == 1]
    else:
        old_gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        p0 = cv.goodFeaturesToTrack(old_gray, mask=None, **feature_params)
        continue;

    for i,(new,old) in enumerate(zip(good_new, good_old)):
        a,b = new.ravel()
        c,d = old.ravel()
        # mask = cv.line(mask, (int(a),int(b)),(int(c),int(d)), color[i].tolist(), 2)
        frame = cv.circle(frame,(int(a),int(b)),5,color[i].tolist(),-1)
    img = frame

    cv.imshow('frame',img)
    k = cv.waitKey(30) & 0xff
    print('key', k)
    while(k != 32):
        break;
    if k == 27:
        break
    # Now update the previous frame and previous points
    old_gray = frame_gray.copy()
    p0 = good_new.reshape(-1,1,2)